### ISSUE TITLE
False positive for SC2154

### ISSUE THREAD

**Issue by kneekoo** [koalaman/shellcheck#2312]
#### For bugs
- Rule Id: SC2154
- My shellcheck version: online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
declare -A iso
iso[cinnamon]="linuxmint-20.2-cinnamon-64bit.iso"
iso[mate]="linuxmint-20.2-mate-64bit.iso"
iso[xfce]="linuxmint-20.2-xfce-64bit.iso"

qemu="qemu-system-x86_64 -m 2G -enable-kvm -soundhw hda -smp 2 -vga virtio"

if [[ $# -eq 1 && -v "iso[$1]" ]]; then
    eval "$qemu -cdrom ${iso[$1]}"
elif [[ $# -eq 2 && $2 -eq "oem" ]]; then
    eval "$qemu -cdrom ${iso[$1]} -hda $2.img"
else
    echo "This is a start script for testing Linux Mint with QEMU"
    echo "Usage: $0 cinnamon|mate|xfce [oem]"
fi
```

#### Here's what shellcheck currently says:
```
Line 11	SC2154: oem is referenced but not assigned.
```

"oem" is just a string. SC2154 is not an issue here.

**Comment by koalaman** [comment]
ShellCheck is right, you are referencing `oem` as a variable:

```
oem=42
[[ oem -eq 42 ]] && echo "True"
```

To compare as a string instead of an integer variable, use `=` instead of `-eq`.

While it's a valid way to reference integer variables, it's also commonly used accidentally like here. Chances are ShellCheck should have a separate warning either for unused variables or in general.

**Comment by TinCanTech** [comment]
Technically `man bash(1)` does not make this distinction, either are valid in `[[`

**Comment by koalaman** [comment]
> Technically `man bash(1)` does not make this distinction, either are valid in `[[`

I'm not sure what you mean. `-eq` and `=` are both valid in `[[ .. ]]`, but the former dereferences `oem` thinking it's a variable name, while the latter treats it as a literal string.

**Comment by kneekoo** [comment]
If they're both valid, a hint should be offered in ShellCheck. I'm sure highly skilled people use it with huge scripts and they probably know how to fix these on the spot. To bash newbies (like me) this makes no sense at all and the [bash documentation doesn't go into details](https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html#Bash-Conditional-Expressions), other than saying "-eq" performs an evaluation of arithmetic expressions, while in practice the script does its job, which is confusing.

Also, why "=" and not "=="? Apparently they're both valid, according to the documentation I mentioned, but in many other languages "=" is assignment and "==" is comparison. What's going on with Bash? :P Sorry for going a bit off-topic but this got a bit weird.

**Comment by koalaman** [comment]
This now triggers a new [[SC2309]](https://github.com/koalaman/shellcheck/wiki/SC2309). Thanks!

To see how this bug affects your script, you can run `./yourscript xfce foobar` and see that it's unable to differentiate between `oem` and `foobar` because both evaluate to `0`. Using `=` instead will correctly show the usage info.

Bash treats `=` and `==` the same in `[ .. ]` and `[[ .. ]]`. POSIX only defines `=` and `[ .. ]`.

**Comment by kneekoo** [comment]
Thank you! :)
